subsys/mgmt/serial_util.c

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright Runtime.io 2018. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+#include <crc16.h>
+#include <misc/byteorder.h>
+#include "mgmt/serial.h"
+#include "base64/base64.h"
+
+static u16_t mcumgr_serial_calc_crc(const u8_t *data, int len)
+{
+	return crc16(data, len, 0x1021, 0, true);
+}
+
+/**
+ * Base64-decodes an mcumgr request.  The source and destination buffers can
+ * overlap.
+ */
+static int mcumgr_serial_decode_req(const u8_t *src, u8_t *dst,
+				    int max_dst_len)
+{
+	if (base64_decode_len((char *)src) > max_dst_len) {
+		return -1;
+	}
+
+	return base64_decode(src, dst);
+}
+
+static int mcumgr_serial_parse_op(const u8_t *buf, int len)
+{
+	u16_t op;
+
+	if (len < sizeof(op)) {
+		return -1;
+	}
+
+	memcpy(&op, buf, sizeof(op));
+	op = sys_be16_to_cpu(op);
+
+	if (op != MCUMGR_SERIAL_HDR_PKT && op != MCUMGR_SERIAL_HDR_FRAG) {
+		return -1;
+	}
+
+	return op;
+}
+
+static int mcumgr_serial_parse_len(struct mcumgr_serial_rx_ctxt *rx_ctxt)
+{
+	u16_t len;
+
+	if (rx_ctxt->raw_off < sizeof(len)) {
+		return -1;
+	}
+
+	memcpy(&len, rx_ctxt->buf, sizeof(len));
+	rx_ctxt->pkt_len = sys_be16_to_cpu(len);
+	return rx_ctxt->pkt_len;
+}
+
+static int mcumgr_serial_decode_frag(struct mcumgr_serial_rx_ctxt *rx_ctxt,
+				     const u8_t *frag)
+{
+	int dec_len;
+
+	dec_len = mcumgr_serial_decode_req(
+		frag,
+		rx_ctxt->buf + rx_ctxt->raw_off,
+		rx_ctxt->buf_size - rx_ctxt->raw_off);
+	if (dec_len == -1) {
+		return -1;
+	}
+
+	rx_ctxt->raw_off += dec_len;
+
+	return 0;
+}
+
+/**
+ * Processes a received mcumgr frame.
+ *
+ * @return                      true if a complete packet was received;
```

Comment:
white spaces

Reply:
That is just the way I tend to write doxygen comments - I indent the `@return` and `@param` descriptions.  I find it easier to read within the source when the right column is aligned.

Should I remove the alignment in all the doxygen comments that I added?

Reply:
In most zephyr files  I see alignment to left, think it is expected. Yo might keep more space if you want.

Reply:
keep as is if you prefer.